File: src/workflows/workflowPlannerEngine.ts
/**
 * Converted from Python: core/workflows/workflow_planner_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { buildRuntimeObjective } from "./objectiveEngine.js";

export function buildWorkflowPlan(objective: any, semantic_runtime: any = null, causality: any = null, application_runtime: any = null): any {
  var runtime_objective: any = buildRuntimeObjective(objective);
  semantic_runtime = py.or2(semantic_runtime, () => ({}));
  var domain: any = py.get(py.get(semantic_runtime, "domain", {}), "domain", "saas");
  var plan_steps: any[] = [];
  var index: any;
  var step: any;
  for ([index, step] of py.enumerate(py.at(runtime_objective, "steps"))) {
    py.listAppend(plan_steps, {"id": `step:${py.toStr(index)}`, "action": step, "runtime": _runtimeForStep(step, domain), "depends_on": ((index > 0) ? `step:${py.toStr(py.sub(index, 1))}` : "")});
  }
  if (py.truthy(application_runtime)) {
    var workflow: any = py.get(application_runtime, "workflow", {});
    var edge: any;
    for (edge of py.iter(py.slice(py.get(workflow, "edges", []), null, 100))) {
      py.listAppend(plan_steps, {"id": `app:${py.toStr(py.len(plan_steps))}`, "action": py.toStr(py.get(edge, "relation", "transition")), "runtime": "application", "depends_on": (py.truthy(plan_steps) ? py.at(py.at(plan_steps, (-1)), "id") : "")});
    }
  }
  if (py.truthy(causality)) {
    var inner: any = py.get(causality, "causality", causality);
    var handoff: any;
    for (handoff of py.iter(py.slice(py.get(py.get(inner, "propagation", {}), "handoffs", []), null, 50))) {
      py.listAppend(plan_steps, {"id": `causal:${py.toStr(py.len(plan_steps))}`, "action": "cross_runtime_handoff", "runtime": py.toStr(py.get(handoff, "to", "native")), "depends_on": (py.truthy(plan_steps) ? py.at(py.at(plan_steps, (-1)), "id") : "")});
    }
  }
  return {"objective": objective, "domain": domain, "steps": plan_steps, "bounded": true};
}
export function _runtimeForStep(step: any, domain: any): any {
  if (py.contains(step, "terminal")) {
    return "terminal";
  }
  if ((py.contains(step, "repository") || py.contains(step, "api"))) {
    return "repository";
  }
  if ((py.contains(step, "infra") || py.eq(domain, "infrastructure"))) {
    return "native";
  }
  if (py.contains(step, "notification")) {
    return "desktop";
  }
  return "browser";
}
export { buildRuntimeObjective };

File: lib/src/workflows/workflow_planner_engine.dart
// Port of core/workflows/workflow_planner_engine.py
import 'objective_engine.dart';

String _runtimeForStep(String step, String domain) {
  if (step.contains('terminal')) {
    return 'terminal';
  }
  if (step.contains('repository') || step.contains('api')) {
    return 'repository';
  }
  if (step.contains('infra') || domain == 'infrastructure') {
    return 'native';
  }
  if (step.contains('notification')) {
    return 'desktop';
  }
  return 'browser';
}

Map<String, dynamic> buildWorkflowPlan(
  String objective, {
  Map<String, dynamic>? semanticRuntime,
  Map<String, dynamic>? causality,
  Map<String, dynamic>? applicationRuntime,
}) {
  final Map<String, dynamic> runtimeObjective =
      buildRuntimeObjective(objective);
  final Map<String, dynamic> semantic = semanticRuntime ?? <String, dynamic>{};
  final dynamic domainOuter = semantic['domain'];
  final String domain = (domainOuter is Map && domainOuter['domain'] is String)
      ? domainOuter['domain'] as String
      : 'saas';

  final List<Map<String, dynamic>> planSteps = <Map<String, dynamic>>[];
  final List<dynamic> objSteps =
      List<dynamic>.from(runtimeObjective['steps'] as List<dynamic>);
  for (int index = 0; index < objSteps.length; index++) {
    final String step = objSteps[index] as String;
    planSteps.add(<String, dynamic>{
      'id': 'step:$index',
      'action': step,
      'runtime': _runtimeForStep(step, domain),
      'depends_on': index > 0 ? 'step:${index - 1}' : '',
    });
  }

  if (applicationRuntime != null && applicationRuntime.isNotEmpty) {
    final dynamic wf = applicationRuntime['workflow'];
    final Map<String, dynamic> workflow =
        wf is Map ? Map<String, dynamic>.from(wf) : <String, dynamic>{};
    final List<dynamic> edges = workflow['edges'] is List
        ? List<dynamic>.from(workflow['edges'] as List<dynamic>)
        : <dynamic>[];
    for (final dynamic edgeRaw in edges.take(100)) {
      final Map<String, dynamic> edge = edgeRaw is Map
          ? Map<String, dynamic>.from(edgeRaw)
          : <String, dynamic>{};
      planSteps.add(<String, dynamic>{
        'id': 'app:${planSteps.length}',
        'action': '${edge['relation'] ?? 'transition'}',
        'runtime': 'application',
        'depends_on':
            planSteps.isNotEmpty ? planSteps.last['id'] as String : '',
      });
    }
  }

  if (causality != null && causality.isNotEmpty) {
    final dynamic innerRaw = causality['causality'];
    final Map<String, dynamic> inner =
        innerRaw is Map ? Map<String, dynamic>.from(innerRaw) : causality;
    final dynamic propRaw = inner['propagation'];
    final Map<String, dynamic> prop = propRaw is Map
        ? Map<String, dynamic>.from(propRaw)
        : <String, dynamic>{};
    final List<dynamic> handoffs = prop['handoffs'] is List
        ? List<dynamic>.from(prop['handoffs'] as List<dynamic>)
        : <dynamic>[];
    for (final dynamic handoffRaw in handoffs.take(50)) {
      final Map<String, dynamic> handoff = handoffRaw is Map
          ? Map<String, dynamic>.from(handoffRaw)
          : <String, dynamic>{};
      planSteps.add(<String, dynamic>{
        'id': 'causal:${planSteps.length}',
        'action': 'cross_runtime_handoff',
        'runtime': '${handoff['to'] ?? 'native'}',
        'depends_on':
            planSteps.isNotEmpty ? planSteps.last['id'] as String : '',
      });
    }
  }

  return <String, dynamic>{
    'objective': objective,
    'domain': domain,
    'steps': planSteps,
    'bounded': true,
  };
}
